End to end ssl

tgmachina opened the pull request:
These changes are in service of turning on SSL internally for JupyterHub (issue #370).

To generate certs, I created a separate, module [Certipy](https://github.com/LLNL/certipy) that wraps pyOpenSsl. Certipy has also been uploaded to PyPI.

To test these changes, I added 3 separate test files that turn on internalSSL via monkey patching. Tests intermittently fail when run together, but pass individually. @minrk had initially suggested making all requests to the hub `public_url`, but I wasn't able to work out a good scheme for that. I set up the 3 separate test files to get these changes through review ASAP, but I'd be happy to iterate on changing these tests.

These changes must be used in conjunction with [changes made to CHP](https://github.com/jupyterhub/configurable-http-proxy/pull/170) in order to work (likely automated github testing will fail because of this). I wasn't entirely sure how to specify a certain CHP version as a requirement.

Happy to iterate on whatever changes you see necessary. Thanks for your help and patience!

minrk commented on jupyterhub/spawner.py:
```diff
@@ -666,6 +673,58 @@ def format_string(self, s):
+            home = user.pw_dir
+
+            # Create dir for user's certs wherever we're starting
+            out_dir = "{home}/.jupyter/certs".format(home=home)
```
Let's call this `.jupyterhub/jupyterhub-certs` just to be clear

tgmachina commented:
For the case where the proxy lives separately from the hub, would it be worthwhile to enable a command like `jupyterhub --generate-config` except to set up the initial certificate authorities/certs that `internal_ssl` uses, for example, something like: `jupyterhub --generate-certs`? That would allow users to get and move the certs they need manually.

As for your assessment of `.move_certs`, that is correct. I have to think about the container case a bit more too, because its also important to get the appropriate domain/ip info into the cert in order for ssl verification to work. That capability exists with the use of `alt_names` at certificate creation, but I am unsure if the info is available or known before container start. I'll think about that some more and work on clarifying this API, good call.

I'm working through these changes now. I think the only thing I'm curious about are the docs. I was thinking of expanding the docstrings for `.create_certs` and `.move_certs` as well as including more notes within the `proxy.md`, `spawner.md` and `websecurity.md` docs. Seem reasonable?

minrk commented:
> something like: jupyterhub --generate-certs?

That's worth a try. I'm not 100% sure what will work best.

> because its also important to get the appropriate domain/ip info into the cert in order for ssl verification to work

oooh, that makes things more difficult because it might be hard/impossible to know this before the server starts in some cases. I'm not quite sure how to deal with that.

> I was thinking of expanding the docstrings for .create_certs and .move_certs as well as including more notes within the proxy.md, spawner.md and websecurity.md docs. Seem reasonable?

Yes, absolutely!

tgmachina commented:
Ok, just as an update, here's what I'm adding/changing/pondering:

* Nearly done with updating the docs, just pending the changes below.
* The `Proxy` class will have reference to its own certs either:
    * Created and signed on Proxy init using the `Hub` CA (basically how its done now).
    * Created by some third-party mechanism (Let's Encrypt, Certipy, self-signed, etc) for which the signing CA file is specified in the proxy config so that trust can be propagated appropriately.
* `Docker`/`Kube` cases w.r.t `internal_ssl`:
    * Swarm/Kube managed deployments: as I understand them, the networking can be set such that containers are connected with `tls`. If containers/pods running the single-user notebooks can't reach each other, this seems like it would have about the same effect as using `internal_ssl`.
    * `Docker` does have a mechanism for specifying an `ip` address prior to launch. This would entail drawing a random, unused `ip` from `Docker's` reserved subnet for container `ips`. I can investigate what enabling that would entail.

tgmachina commented:
Ok, I've updated CHP and JupyterHub accordingly. Overview of the changes:
* External proxy certs can be specified via config (for when the proxy is externally managed)
* CHP has two CAs: one for the API and the other for all client connections
* Trust is set up via Certipy's "trust_graph". Basically what component trusts what can be specified in a more human readable fashion.

Sorry for the delay in getting that out. I'll be a bit busy over the coming weeks, but I'll keep an eye out as always and look for anything else that needs changing/adding.

minrk commented:
Terrific! I'll try to get a review of this soon

minrk commented:
For the docker networking case, I don't think the pre-chosen ip ought to be an issue because the container name also works as a connectable hostname within docker networks, and that's known and chosen by us prior to start.

That should leave docker with the following cases:

1. internal docker (use container name as connect hostname instead of ip)
2. docker-exposed network (static public hostname/ip of machine)
3. exposed localhost docker (use 127.0.0.1)

I'm going to try to write a PR to dockerspawner to get a handle on this.

In the meantime, what would be a super useful thing to add is some tests that verify that various of these connections are actually rejected when requests aren't made with the right certs. I don't see those in the PR, unless I'm missing something.

minrk commented:
@tgmachina is it alright if I update this PR with changes as I test? I might want to tweak a few things, but I don't want to conflict with work you are doing.

minrk commented:
I've been able to get a version of this running with DockerSpawner with the proxy local to the hub. I had to make one adjustment to get it to work:

- make `move_certs` async. That's a minor change, but it requires moving the call to a different stage, since `.get_env` shouldn't have side effects, it should only report environment variables based on already-created state.

The next hiccup I'm hitting is running the proxy separate from the Hub, which is a best practice in container deployments. This is tricky right now due to the order of SSL initialization. I think setting up the proxy certs needs to happen in `.init_internal_ssl` rather than `init_proxy`, because `jupyterhub --generate-certs` needs to result in creating all the certs required.

tgmachina commented:
Hey @minrk! Got it, so `.create_certs` and `.move_certs` should be in their own phase outside of `.get_env`. Was that the change you were going to send back? Feel free to push any changes that you need to, no problem. I've been revamping the tests and working on adding more to test that connections fail as you asked.

tgmachina commented:
My altering of the app fixture to respond to `ssl_enabled` in tests was flawed. I've fixed the fixture and updated the tests accordingly. The main change was ensuring that when a test required making a request to a component, that the correct certs were attached when the `ssl_enabled` flag is turned on.

In addition to fixing the existing tests, I added a `test_internal_ssl_connections.py` file that has tests to ensure that connections to components with `internal_ssl` enabled, fail without the right certs.

minrk commented:
Okay, after resolving the subdomain issues with trusted_alt_names, I believe this is totally working. I just need to update the tests with real-database backends, because `test_api` can't actually be run multiple times against the same database.

consideRatio commented:
Wooooooooooooo wow this PR! Massive work done here! I've been keeping an eye on it, well done @tgmachina and @minrk !!! :tada:

tgmachina commented:
Wow! Thanks so much for all your help @minrk!

tgamblin commented:
This is awesome!  Thanks @tgmachina and @minrk!
